QRDA Import date fixes and class fixes

sjpadgett opened the pull request:
fix careplan dates to return null when empty
fix negate for medication order
fix negated helper logic

sjpadgett commented:
@kchapple note.

Once I figure out why cypress has me download a patient test deck of patients for import for a given measure and yet errors when I give the same patients back for test i.e gives 78 patients then says only 75 expected, we'll get green! 
However for the CMS22v10 we pass validation and calculations, yeah. May not seem like much but for us, a big deal!

![image](https://user-images.githubusercontent.com/1859985/163306066-1d53663a-9591-4a21-8e2e-5e9469c9415a.png)

bradymiller commented:
![2DV-2](https://user-images.githubusercontent.com/278968/163309085-670a16a4-d038-41ee-be66-91e3830e6465.gif)

stephenwaite commented:
@sjpadgett , just let me know when you'd like me to test something please. I created a login on your cypress.

sjpadgett commented:
I'll send you the test steps in email then follow the c(1) where you have to enter attributes into cypress form.
The test run like a check list. I haven't looked to closely at the test so it's an adventure we're both on!:)

sjpadgett commented:
I'll set you up a product to use. Give me a few minutes

sjpadgett commented:
@stephenwaite Set you up product Stephen W CQM

sjpadgett commented:
Okay thanks but I'm really confused where the device template went. I rechecked cat3 and cat1 so should be good.

sjpadgett commented:
sorry @kchapple missed concern on last commit. tired!

![image](https://user-images.githubusercontent.com/1859985/163657212-f47a7d83-3e53-4b57-898b-3a299c3765c2.png)

sjpadgett commented:
@kchapple Decided to go a different way with encounter ends date. It gets blurry using calendar events relationship to encounter.
I added end date to form_encounter and modified the qdm encounter service to reflect.

sjpadgett commented:
@kchapple Need service for careplan device_order

![image](https://user-images.githubusercontent.com/1859985/163693976-63be2468-d04c-44db-8cf3-0718e664d5cb.png)

sjpadgett commented:
have to add reason to procedure orders!!

sjpadgett commented:
@kchapple I had to add reason code to procedure intervention preformed type that the ProcedureService uses.
Can you add reason to the service. I would add negation also on reason status though I haven't found any to this template in any patient samples, it is in the spec.

sjpadgett commented:
Never mind on reason I was able to do it. Just check my commit for me to ensure I don't step on anything

sjpadgett commented:
@kchapple This makes CMS22v10, CMS69v10 and CMS125v10 is a go. Next for me will be CMS122v10 but hooray for us.
Will merge this PR and start again with next testing hiccup!

sjpadgett commented:
I've learned that we are expected to validate imports! nuts, how! Anyway using normal test where invalid reports slip by me I can get calculator errors where the report shouldn't have been submitted. So until I figure out I have to validate imports I am testing against cypress known good patients. e.g. Here is CMS122v10 normal test(this patient didn't have any encounters):

![image](https://user-images.githubusercontent.com/1859985/163739110-f9787792-6c75-4029-9c4a-a7fa25bd3643.png)

And with cypress known good patients:

![image](https://user-images.githubusercontent.com/1859985/163738908-1cdbc652-80f9-422f-998f-56790d850c46.png)

So this is good news and bad news! Anyway, add CMS122v10 to list.

sjpadgett commented:
And CMS127v10

![image](https://user-images.githubusercontent.com/1859985/163741369-7b6c87f2-339f-44fd-8a29-c3bb84ad4b09.png)

sjpadgett commented:
CMS130v10 Anyone getting bored seeing green?

![image](https://user-images.githubusercontent.com/1859985/163748544-84b45b97-a131-4aaa-8b34-bd3b9e8d5a45.png)

bradymiller commented:
go green!!!

https://www.youtube.com/watch?v=BVDnDORSY9E

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -315,3 +315,16 @@ ALTER TABLE `form_care_plan` CHANGE `date` `date` DATETIME NULL DEFAULT NULL;
+#IfMissingColumn procedure_order_code date_end
+ALTER TABLE `procedure_order_code` ADD `date_end` datetime DEFAULT NULL;
+ALTER TABLE `procedure_order_code` ADD `reason_code` varchar(31) DEFAULT NULL;
+ALTER TABLE `procedure_order_code` ADD `reason_description` text DEFAULT NULL;
```
remove the default null from text (here and in database.sql)

sjpadgett commented:
Needs fixed in service. 

![image](https://user-images.githubusercontent.com/1859985/163829108-42c2e128-e24c-4f21-bec5-24f349cfa325.png)

sjpadgett commented:
@kchapple be sure to run sql upgrade procedure code size was to small for an oid

sjpadgett commented:
@kchapple you'll need to add negation to immunization service while I go fix conflicts, again! I added reason code and status to table.

sjpadgett commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -316,6 +316,203 @@ ALTER TABLE `form_care_plan` CHANGE `date` `date` DATETIME NULL DEFAULT NULL;
+START TRANSACTION;
+INSERT INTO lang_updates_610 VALUES
+('abk','ab'),
+('aar','aa'),
```
@bradymiller @adunsulag I had to fix conflicts so you guys need to ensure sql upgrade and database are still intact.
Looks okay to me but check me anyways. Thanks

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -511,4 +511,32 @@ DROP TABLE lang_updates_610;
+ALTER TABLE `procedure_order` CHANGE `date_ordered` `date_ordered` DATETIME DEFAULT NULL;
+#EndIf
+
+#IfMissingColumn immunizations reason_code
```
immunizations changes are missing in database.sql

sjpadgett replied:
thanks

bradymiller commented on sql/database.sql:
```diff
@@ -9236,14 +9237,20 @@ DROP TABLE IF EXISTS `procedure_order_code`;
+  `reason_description` text,
+  `reason_date_low` datetime DEFAULT NULL,
+  `reason_date_high` datetime DEFAULT NULL,
+  `reason_status` varchar(32) DEFAULT NULL,
```
varchar is 31 in upgrade

sjpadgett commented:
gesh! what a marathon!
@kchapple immunization template is not seeing the oid valueset and sets negationInd to 'false'
The nulled cvx code is in diff component than we're used to seeing.

Will bring in so we can reset devs to see why you can't get diagnosis to populate issue_encounter. Maybe change to windows! lol
